internal/cluster-agent/agent.go

```diff
@@ -0,0 +1,274 @@
+			a.closeConnection()
+			return ctx.Err()
+		case <-a.stopChan:
+			a.logger.Info("agent stopping")
+			a.closeConnection()
+			return nil
+		default:
+			if err := a.connect(); err != nil {
+				a.logger.Error("connection failed",
+					"error", err,
+					"retryAfter", a.config.ReconnectDelay,
+				)
+
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case <-time.After(a.config.ReconnectDelay):
+					continue
+				}
+			}
+
+			a.handleConnection(ctx)
+
+			a.logger.Info("connection lost, reconnecting",
+				"delay", a.config.ReconnectDelay,
+			)
+
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-time.After(a.config.ReconnectDelay):
+				continue
+			}
+		}
+	}
+}
+
+func (a *Agent) Stop() {
+	close(a.stopChan)
+}
+
+// connect establishes a WebSocket connection to the control plane
+func (a *Agent) connect() error {
+	u, err := url.Parse(a.config.ServerURL)
+	if err != nil {
+		return fmt.Errorf("invalid server URL: %w", err)
+	}
+
+	query := u.Query()
+	query.Set("plane", a.config.PlaneName)
+	query.Set("cluster", a.config.ClusterName)
+	u.RawQuery = query.Encode()
+
+	tlsConfig := &tls.Config{
+		Certificates:       []tls.Certificate{a.clientCert},
+		RootCAs:            a.serverCA,
+		MinVersion:         tls.VersionTLS12,
+		InsecureSkipVerify: a.serverCA == nil, //nolint:gosec // Intentional: insecure only when no CA provided (dev mode)
+	}
+
+	dialer := websocket.Dialer{
+		TLSClientConfig:  tlsConfig,
+		HandshakeTimeout: 10 * time.Second,
+	}
+
+	a.logger.Info("connecting to control plane", "url", u.String())
+
+	conn, _, err := dialer.Dial(u.String(), nil)
+	if err != nil {
+		return fmt.Errorf("dial failed: %w", err)
+	}
+
+	a.mu.Lock()
+	a.conn = conn
+	a.mu.Unlock()
+
+	a.logger.Info("connected to control plane")
+	return nil
+}
+
+// handleConnection handles an established WebSocket connection
+func (a *Agent) handleConnection(ctx context.Context) {
+	// Setup ping/pong handlers for connection health
+	a.conn.SetPingHandler(func(appData string) error {
+		a.logger.Debug("received ping from server")
+		return a.conn.WriteControl(websocket.PongMessage, []byte(appData), time.Now().Add(time.Second))
+	})
+
+	// Read messages
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			_, message, err := a.conn.ReadMessage()
+			if err != nil {
+				if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway, websocket.CloseAbnormalClosure) {
+					a.logger.Error("websocket error", "error", err)
+				}
+				return
+			}
+
+			// Parse DispatchableRequest
+			var dispatchableReq messaging.DispatchableRequest
+			if err := json.Unmarshal(message, &dispatchableReq); err != nil {
+				a.logger.Warn("failed to parse request", "error", err)
+				continue
+			}
+
+			// Validate request has ID
+			if dispatchableReq.RequestID == "" {
+				a.logger.Warn("received request without requestID")
+				continue
+			}
+
+			// Handle request
+			a.handleDispatchableRequest(ctx, &dispatchableReq)
```

Comment:
this blocks the read loop until handleRequest returns.

Might be as simple of a fix as `go a.handleRequest()` but see if there's any consequences to doing this